backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/CallableReferenceLowering.kt

```diff
@@ -183,48 +183,64 @@ internal class CallableReferenceLowering(val context: Context): FileLoweringPass
-                lastParameterType as IrSimpleType
-                // If the last parameter is Continuation<> inherit from SuspendFunction.
-                suspendFunctionIrClass = symbols.suspendFunctions[numberOfParameters - 1].owner
-                val suspendFunctionClassTypeParameters = functionParameterTypes.dropLast(1) +
-                        (lastParameterType.arguments.single().typeOrNull ?: irBuiltIns.anyNType)
-                superTypes += suspendFunctionIrClass.symbol.typeWith(suspendFunctionClassTypeParameters)
+            val superTypes = mutableListOf<IrType>()
+            val functionClass: IrClass
+            val suspendFunctionClass: IrClass?
+            when {
+                isKFunction -> {
+                    superTypes += kFunctionImplSymbol.typeWith(referencedFunction.returnType)
+                    functionClass = symbols.kFunctions[numberOfParameters].owner
+                    superTypes += functionClass.typeWith(functionParameterTypes + referencedFunction.returnType)
+                    val lastParameterType = unboundFunctionParameters.lastOrNull()?.type
+                    if (lastParameterType?.classifierOrNull != continuationClassSymbol)
+                        suspendFunctionClass = null
+                    else {
+                        lastParameterType as IrSimpleType
+                        // If the last parameter is Continuation<> inherit from SuspendFunction.
+                        suspendFunctionClass = symbols.suspendFunctions[numberOfParameters - 1].owner
+                        val suspendFunctionClassTypeParameters = functionParameterTypes.dropLast(1) +
+                                (lastParameterType.arguments.single().typeOrNull ?: irBuiltIns.anyNType)
+                        superTypes += suspendFunctionClass.symbol.typeWith(suspendFunctionClassTypeParameters)
+                    }
+                }
+                isKSuspendFunction -> {
+                    superTypes += kSuspendFunctionImplSymbol.typeWith(referencedFunction.returnType)
+                    functionClass = symbols.functions[numberOfParameters + 1].owner
+                    val continuationType = continuationClassSymbol.typeWith(referencedFunction.returnType)
+                    superTypes += functionClass.typeWith(functionParameterTypes + continuationType + irBuiltIns.anyNType)
+                    suspendFunctionClass = symbols.kSuspendFunctions[numberOfParameters].owner
+                    superTypes += suspendFunctionClass.typeWith(functionParameterTypes + referencedFunction.returnType)
+                }
+                else -> {
+                    superTypes += irBuiltIns.anyType
+                    functionClass = symbols.functions[numberOfParameters].owner
+                    superTypes += functionClass.typeWith(functionParameterTypes + referencedFunction.returnType)
+                    // Anonymous functions and lambdas never inherit SuspendFunctionX.
```

Comment:
Why so?

Reply:
Suspend lambdas have type `SuspendLambdaX`, not `KSuspendLambdaX`. So only lambdas that can be at this point are not suspend.

Reply:
AFAIR, the code with `suspendFunctionIrClass ` was intended to make this work:
```
import kotlin.coroutines.*

fun foo(block: (Continuation<Unit>) -> Any?) {
    block as (suspend () -> Unit)
}

fun main() {
    foo {}
}
```

Isn't the test broken now?

Reply:
You're right, I for some reason only supported that for KFunction.

Reply:
Wouldn't say ` block as (suspend () -> Unit)` must work, although it is nice to have.